src/Concordium/Client/GRPC2.hs

```diff
@@ -0,0 +1,2658 @@
+    Word64
+deMkWord64 val = coerce $ val ^. ProtoFields.value
+
+-- |Like 'deMkWord64' above, but for Word32 fields instead.
+deMkWord32 ::
+    ( Data.ProtoLens.Field.HasField
+        b
+        "value"
+        Word32
+    ) =>
+    b ->
+    Word32
+deMkWord32 val = coerce $ val ^. ProtoFields.value
+
+-- |Like 'deMkWord64' above, but for Word16 fields instead.
+deMkWord16 ::
+    ( Data.ProtoLens.Field.HasField
+        b
+        "value"
+        Word32
+    ) =>
+    b ->
+    Maybe Word16
+deMkWord16 val =
+    if v <= fromIntegral (maxBound :: Word16)
+        then return $ fromIntegral v
+        else Nothing
+  where
+    v = val ^. ProtoFields.value
+
+-- |Like 'mkWord32', but for Word8 fields instead.
+deMkWord8 ::
+    ( Data.ProtoLens.Field.HasField
+        b
+        "value"
+        Word32
+    ) =>
+    b ->
+    Maybe Word8
+deMkWord8 val =
+    if v <= fromIntegral (maxBound :: Word8)
+        then return $ fromIntegral v
+        else Nothing
+  where
+    v = val ^. ProtoFields.value
+
+-- |Validate a protocol buffer message with an `url` Text field.
+-- Returns @Nothing@ if the string is too long and the text
+-- wrapped in @UrlText@ otherwise.
+deMkUrlText ::
+    ( Data.ProtoLens.Field.HasField
+        b
+        "url"
+        Text
+    ) =>
+    b ->
+    Maybe UrlText
+deMkUrlText val =
+    if Text.length v <= fromIntegral maxUrlTextLength
+        then return $ UrlText v
+        else Nothing
+  where
+    v = val ^. ProtoFields.url
+
+-- |Decode a bytestring and return the decoded value if the
+-- input was consumed. Returns @Nothing@ if the bytestring
+-- could not be decoded or if the input was not consumed.
+decodeAndConsume :: (S.Serialize a) => ByteString -> Maybe a
+decodeAndConsume bs = do
+    case S.runGet getter bs of
+        Left _ -> Nothing
+        Right v -> v
+  where
+    getter = do
+        v <- S.get
+        empty <- S.isEmpty
+        unless empty $ fail "decodeAndConsume did not consume its input."
+        return v
+
+-- |A helper class analogous to something like Aeson's FromJSON.
+-- It exists to make it more manageable to convert the Protobuf
+-- types to their internal Haskell type equivalents.
+class FromProto a where
+    -- |The corresponding Haskell type.
+    type Output a
+
+    -- |A conversion function from the protobuf type to its Haskell
+    -- equivalent. Returns Nothing if the conversion failed.
+    fromProto :: a -> Maybe (Output a)
+
+instance FromProto Proto.AccountIndex where
+    type Output Proto.AccountIndex = AccountIndex
+    fromProto = return . AccountIndex . deMkWord64
+
+instance FromProto Proto.AccountAddress where
+    type Output Proto.AccountAddress = AccountAddress
+    fromProto = deMkSerialize
+
+instance FromProto Proto.CredentialRegistrationId where
+    type Output Proto.CredentialRegistrationId = (CredentialRegistrationID, RawCredentialRegistrationID)
+    fromProto crId = do
+        raw <- deMkSerialize crId
+        nonRaw <- credIdFromRaw raw
```

Comment:
Where is this function defined? There is no PR in base, nor is this in main there.

Reply:
It is in a feature branch, you can click the folder in GitHub to see the commit ref, or it is here:

https://github.com/Concordium/concordium-base/tree/d132f5de85753b3ac7200c60ba5f756eef270dd3

So I guess we should also have have PR’s for things that are not PRs that pull into main then, right?

Reply:
Here, specifically: https://github.com/Concordium/concordium-base/blob/d132f5de85753b3ac7200c60ba5f756eef270dd3/haskell-src/Concordium/ID/Types.hs#L952

Reply:
Right, but that is not merged to main, right? Nor is there a PR? Or where is this?